lib/supavisor.ex

```diff
@@ -156,79 +168,100 @@ defmodule Supavisor do
   ## Internal functions
 
   @spec start_local_pool(id, secrets) :: {:ok, pid} | {:error, any}
-  defp start_local_pool({tenant, user, mode} = id, {method, secrets}) do
-    Logger.debug("Starting pool for #{inspect(id)}")
+  defp start_local_pool({{type, tenant}, _user, _mode} = id, secrets) do
+    Logger.warn("Starting pool(s) for #{inspect(id)}")
 
-    case Tenants.get_pool_config(tenant, secrets.().alias) do
-      %Tenant{} = tenant_record ->
-        %{
-          db_host: db_host,
-          db_port: db_port,
-          db_database: db_database,
-          default_parameter_status: ps,
-          ip_version: ip_ver,
-          default_pool_size: def_pool_size,
-          default_max_clients: def_max_clients,
-          users: [
-            %{
-              db_user: db_user,
-              db_password: db_pass,
-              pool_size: pool_size,
-              # mode_type: mode_type,
-              max_clients: max_clients
-            }
-          ]
-        } = tenant_record
-
-        {pool_size, max_clients} =
-          if method == :auth_query do
-            {def_pool_size, def_max_clients}
-          else
-            {pool_size, max_clients}
-          end
-
-        auth = %{
-          host: String.to_charlist(db_host),
-          port: db_port,
-          user: db_user,
-          database: db_database,
-          password: fn -> db_pass end,
-          application_name: "supavisor",
-          ip_version: H.ip_version(ip_ver, db_host),
-          upstream_ssl: tenant_record.upstream_ssl,
-          upstream_verify: tenant_record.upstream_verify,
-          upstream_tls_ca: H.upstream_cert(tenant_record.upstream_tls_ca),
-          require_user: tenant_record.require_user,
-          secrets: secrets
-        }
+    user = elem(secrets, 1).().alias
 
-        args = %{
-          id: id,
-          tenant: tenant,
-          user: user,
-          auth: auth,
-          pool_size: pool_size,
-          mode: mode,
-          default_parameter_status: ps,
-          max_clients: max_clients
-        }
+    case type do
+      :single -> T.get_pool_config(tenant, user)
+      :cluster -> T.get_cluster_config(tenant, user)
+    end
+    |> case do
+      [_ | _] = replicas ->
+        opts =
+          Enum.map(replicas, fn replica ->
+            # cluster config has "type" field, pool config doesn't
+            if Map.has_key?(replica, :type) do
```

Comment:
I would suggest to pattern match on the type instead. It is probably more reliable than checking on a field;

```
case replica do
  %ClusterTenant{tenant: tenant, type: type} -> Map.put(tenant, :replica_type, type)
  %Tenant{...} = tenant -> Map.put(tenant, :replica_type, :write)
```

Alternatively make it so both get_cluster_config and get_pool_config return the same data structure, already pre-normalized.